# function that takes a MIDORI2 RDP-format reference database and extracts amplicons

extract_amplicons_crabs <- function(ref_seqs = taxon_refdb, 
                                    fwd, rev, l, L,
                                   out,
                                   conda_dir = "/Users/mikea/miniconda3/bin/conda",
                                   conda_env = "crb2",
                                   crabs_path = "crabs", # or "python ~/path/to/reference_database_creator/crabs" to use the newest version
                                   tax_path = "/Users/mikea/Documents/mikedata/GhostBLASTer/BLASTrees/tax_20250609/",
                                   # all_starts = FALSE,
                                   verbose = TRUE){
  
  # make working directory if needed
  if (!dir.exists(out)) {
    dir.create(out)
  }
  
  # detect if imported crabs format exists already
  imported_already <- file.exists(paste0(out, "crabs_raw_seqs.txt"))
  
  # set up code for all start positions (not working due to crabs bug at the moment)
    # (i.e., not restricted to include primer matches when true)
  # if(all_starts == TRUE){rev2 <- paste0(rev, " --all-start-positions")}else{rev2 <- rev}
  
  if(imported_already == FALSE){
    
  if (verbose == TRUE) {message("Loading reference sequences into CRABS...")}
  # detect if refseqs is a fasta file and if not treat it like a DNAStringset
  if(grepl(ref_seqs[1], pattern = "fasta")){
    ref_seqs_path <- ref_seqs
  }else{
    writeXStringSet(ref_seqs, paste0(out, "raw_seqs.fasta"))
    ref_seqs_path <- paste0(out, "raw_seqs.fasta")
  }

  if (verbose == TRUE) {tictoc::tic("Import complete...")}
system2(conda_dir, 
        args = c("run", "-n", conda_env, crabs_path, 
                 "--import",
                 "--import-format", "NCBI",
                 "--input", ref_seqs_path,
                 "--names", paste0(tax_path, "names.dmp"),
                 "--nodes", paste0(tax_path, "nodes.dmp"),
                 "--acc2tax", paste0(tax_path, "nucl_gb.accession2taxid"),
                 "--output", paste0(out, "crabs_raw_seqs.txt"),
                 "--ranks", "'superkingdom;phylum;class;order;family;genus;species'"
        ), 
        stdout = TRUE, stderr = TRUE)
if (verbose == TRUE) {tictoc::toc()}
}else{message("crabs_raw_seqs.txt file detected. Skipping import step ...")}

if (verbose == TRUE) {message("Performing in silico PCR ...")}
# In silico PCR to extract amplicons with primer regions (w/ â‰¤ 4 mismatches)
if (verbose == TRUE) {tictoc::tic("PCR complete ...")}
system2(conda_dir, 
        args = c("run", "-n", conda_env, crabs_path, 
                 "--in-silico-pcr",
                 "--input", paste0(out, "crabs_raw_seqs.txt"),
                 "--output", paste0(out, "crabs_amplicons_pcr.txt"),
                 "--forward", fwd,
                 "--reverse", rev), 
        stdout = TRUE, stderr = TRUE)
if (verbose == TRUE) {tictoc::toc()}

if (verbose == TRUE) {message("Performing pga to extract amplicons without primer regions ...")}
if (verbose == TRUE) {tictoc::tic("pga complete ...")}
pga_arg_vector <- c("run", "-n", conda_env, crabs_path, 
                    "--pairwise-global-alignment",
                    "--input", paste0(out, "crabs_raw_seqs.txt"),
                    "--amplicons", paste0(out, "crabs_amplicons_pcr.txt"),
                    "--output", paste0(out, "crabs_amplicons_pga.txt"),
                    "--forward", fwd,
                    "--reverse", rev,
                    "--size-select", "20000", 
                    "--percent-identity", "0.70",
                    "--coverage", "0.95")
system2(conda_dir, 
        args = pga_arg_vector, 
        stdout = TRUE, stderr = TRUE)
if (verbose == TRUE) {tictoc::toc()}

# message(paste(pga_arg_vector, sep = " "))
}
